rust/src/storage/s3.rs

```diff
@@ -14,7 +15,7 @@ use super::{parse_uri, ObjectMeta, StorageBackend, StorageError};
 #[derive(Debug, PartialEq)]
 pub struct S3Object<'a> {
     pub bucket: &'a str,
-    pub key: &'a str,
+    pub key: Cow<'a, str>,
```

Comment:
Instead of creating a new String and hold a reference to it through Cow, it is more idiomatic to just use the String type here.

Alternatively, we could decide to not handle redundant cases like `s3://foo///bar`, which is the case for AWS cli. The we can keep using `&'a str` here by taking a reference to the original `&str` through https://doc.rust-lang.org/std/primitive.str.html#method.trim_left_matches.

Reply:
I could keep the redundant test case and refactoring the code with the String type. Or do you prefer to keep the `&'a str`, delete the redundant test case and find another solution for removing the `"//"`?

Reply:
I am leaning more towards not handling redundant slashes in delta.rs and leave it to the AWS S3 SDK.

Reply:
Maybe we could just use https://doc.rust-lang.org/std/primitive.str.html#method.trim_end_matches for `log_path` variable when creating the DeltaTable?

Reply:
The original issue that I filed @houqp  was around _trailing_ slashes in #62. I don't know how the Delta bindings for Scala treat this, but it's silly to me that `s3://delta/simple` is a table but `s3://delta/simple/` is not.

Reply:
@fvaleye my bad for posting the wrong link in my previous comment, trim_end_matches is the the api to use here.

Reply:
I added `trim_end_matches` when creating the DeltaTable to keep using `&'a str` in the S3Bucket and leave the S3 URI to AWS S3 SDK. The error was raised in S3Storage because the checkpoint couldn't be read because of "s3://delta/simple//_delta_log" S3 URI.

Reply:
The path normalization needs to happen at the beginning of https://github.com/delta-io/delta-rs/blob/main/rust/src/delta.rs#L526. Ideally it needs to be done in a platform agnostic way based on storage backends at use, but we can leave that refactoring to https://github.com/delta-io/delta-rs/issues/40.

Reply:
I will look into the #40 for level up in Rust :)

Reply:
That would be great, while you work on that, i recommend also changing the code to do the path normalization at the beginning of `DeltaTable::new` function so `table_path: table_path.to_string()` also gets the consistent value without trailing slash.